core/src/layers/minitrace.rs

```diff
@@ -0,0 +1,349 @@
+// under the License.
+
+use std::fmt::Debug;
+use std::io;
+use std::task::Context;
+use std::task::Poll;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::FutureExt;
+use minitrace::prelude::*;
+
+use crate::ops::*;
+use crate::raw::*;
+use crate::*;
+
+/// Add [minitrace](https://docs.rs/minitrace/) for every operations.
+///
+/// # Examples
+///
+/// ## Basic Setup
+///
+/// ```
+/// use anyhow::Result;
+/// use opendal::layers::MinitraceLayer;
+/// use opendal::services;
+/// use opendal::Operator;
+///
+/// let _ = Operator::new(services::Memory::default())
+///     .expect("must init")
+///     .layer(MinitraceLayer)
+///     .finish();
+/// ```
+///
+/// ## Real usage
+///
+/// ```no_run
+/// use std::error::Error;
+///
+/// use anyhow::Result;
+/// use futures::executor::block_on;
+/// use opendal::layers::MinitraceLayer;
+/// use opendal::services;
+/// use opendal::Operator;
+///
+/// fn main() -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+///     let collector = {
+///         let (span, collector) = minitrace::Span::root("op");
+///         let _g = span.set_local_parent();
+///         let runtime = tokio::runtime::Runtime::new()?;
+///
+///         runtime.block_on(async {
+///             let _ = dotenvy::dotenv();
+///             let op = Operator::from_env::<services::Memory>()
+///                 .expect("init operator must succeed")
+///                 .layer(MinitraceLayer)
+///                 .finish();
+///
+///             op.write("test", "0".repeat(16 * 1024 * 1024).into_bytes())
+///                 .await
+///                 .expect("must succeed");
+///             op.stat("test").await.expect("must succeed");
+///             op.read("test").await.expect("must succeed");
+///         });
+///         collector
+///     };
+///
+///     let spans = block_on(collector.collect());
+///
+///     let bytes =
+///         minitrace_jaeger::encode("opendal".to_owned(), rand::random(), 0, 0, &spans).unwrap();
+///     minitrace_jaeger::report_blocking("127.0.0.1:6831".parse().unwrap(), &bytes)
+///         .expect("report error");
+///
+///     Ok(())
+/// }
+/// ```
+///
+/// # Output
+///
+/// OpenDAL is using [`minitrace`](https://docs.rs/minitrace/latest/minitrace/) for tracing internally.
+///
+/// To enable minitrace output, please init one of the reporter that `minitrace` supports.
+///
+/// For example:
+///
+/// ```ignore
+/// extern crate minitrace_jaeger;
+///
+/// let spans = block_on(collector.collect());
+///
+/// let bytes =
+///     minitrace_jaeger::encode("opendal".to_owned(), rand::random(), 0, 0, &spans).unwrap();
+/// minitrace_jaeger::report_blocking("127.0.0.1:6831".parse().unwrap(), &bytes).expect("report error");
+/// ```
+///
+/// For real-world usage, please take a look at [`minitrace-datadog`](https://crates.io/crates/minitrace-datadog) or [`minitrace-jaeger`](https://crates.io/crates/minitrace-jaeger) .
+pub struct MinitraceLayer;
```

Comment:
It's better to use `MinitraceLayer {}` and derive `Default` for it so that we can add new config value for it without breaking.

Reply:
This is close to TracingLayer and MetricsLayer for now, and if we need to refactor, I may do it in other PRs.